File: scripts/add_page.py
#!/usr/bin/env python3
"""
Script to create a new page in the SvelteKit routes directory.
Uses templates from templates/page/ to generate the page files.
"""

from pathlib import Path

import argparse
import sys


def create_page(page_path: str, template_dir: Path, routes_dir: Path) -> None:
    """Create a new page from templates."""
    # Normalize the page path (remove leading/trailing slashes)
    page_path = page_path.strip("/")

    # Create the target directory
    target_dir = routes_dir / page_path
    target_dir.mkdir(parents=True, exist_ok=True)

    # Copy template files
    template_files = ["+layout.svelte", "+page.server.ts", "+page.svelte", "+page.ts"]

    for template_file in template_files:
        template_path = template_dir / template_file
        if template_path.exists():
            target_path = target_dir / template_file

            # Read template content
            with open(template_path, "r", encoding="utf-8") as f:
                content = f.read()

            # Write to target
            with open(target_path, "w", encoding="utf-8") as f:
                f.write(content)

            print(f"Created: {target_path}")
        else:
            print(f"Warning: Template {template_file} not found")


def main():
    parser = argparse.ArgumentParser(
        description="Create a new SvelteKit page from templates"
    )
    parser.add_argument(
        "path", help='Path for the new page (e.g., "about" or "blog/post")'
    )

    args = parser.parse_args()

    # Get script directory and project root
    script_dir = Path(__file__).parent
    project_root = script_dir.parent

    # Define paths
    template_dir = project_root / "templates" / "page"
    routes_dir = project_root / "src" / "routes"

    # Validate paths exist
    if not template_dir.exists():
        print(f"Error: Template directory not found: {template_dir}")
        sys.exit(1)

    if not routes_dir.exists():
        print(f"Error: Routes directory not found: {routes_dir}")
        sys.exit(1)

    try:
        create_page(args.path, template_dir, routes_dir)
        print(f"\nPage created successfully at: src/routes/{args.path}")
    except Exception as e:
        print(f"Error creating page: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: tools/add-page.ts
#!/usr/bin/env node
/**
 * Script to create a new page in the SvelteKit routes directory.
 * Uses templates from templates/page/ to generate the page files.
 */

import { mkdir, readFile, writeFile } from "fs/promises"
import { existsSync } from "fs"
import { join, dirname } from "path"
import { fileURLToPath } from "url"

const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

async function createPage(pagePath: string, templateDir: string, routesDir: string): Promise<void> {
	// Normalize the page path (remove leading/trailing slashes)
	pagePath = pagePath.replace(/^\/+|\/+$/g, "")

	// Create the target directory
	const targetDir = join(routesDir, pagePath)
	await mkdir(targetDir, { recursive: true })

	// Copy template files
	const templateFiles = ["+layout.svelte", "+page.server.ts", "+page.svelte"]

	for (const templateFile of templateFiles) {
		const templatePath = join(templateDir, templateFile)
		if (existsSync(templatePath)) {
			const targetPath = join(targetDir, templateFile)

			// Read template content
			const content = await readFile(templatePath, "utf-8")

			// Write to target
			await writeFile(targetPath, content)

			console.log(`Created: ${targetPath}`)
		} else {
			console.log(`Warning: Template ${templateFile} not found`)
		}
	}
}

async function main() {
	const args = process.argv.slice(2)
	if (args.length === 0) {
		console.log("Usage: node add_page.ts <path>")
		console.log("Example: node add_page.ts about")
		console.log("Example: node add_page.ts blog/post")
		process.exit(1)
	}

	const pagePath = args[0]

	// Get script directory and project root
	const scriptDir = __dirname
	const projectRoot = dirname(scriptDir)

	// Define paths
	const templateDir = join(projectRoot, "templates", "page")
	const routesDir = join(projectRoot, "src", "routes")

	// Validate paths exist
	if (!existsSync(templateDir)) {
		console.error(`Error: Template directory not found: ${templateDir}`)
		process.exit(1)
	}

	if (!existsSync(routesDir)) {
		console.error(`Error: Routes directory not found: ${routesDir}`)
		process.exit(1)
	}

	try {
		await createPage(pagePath, templateDir, routesDir)
		console.log(`\nPage created successfully at: src/routes/${pagePath}`)
	} catch (error) {
		console.error(`Error creating page: ${error}`)
		process.exit(1)
	}
}

if (import.meta.url === `file://${process.argv[1]}`) {
	main()
}
